server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.get;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TransportGetFromTranslogAction extends HandledTransportAction<GetRequest, TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, GetRequest::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetRequest request, ActionListener<Response> listener) {
+        final ShardId shardId = request.getInternalShardId();
```

Comment:
I've added a new Request type that is just a wrapper basically. But since `GetRequest` itself doesn't implement any equality/etc, I haven't done it for this either. Not sure if it worths all the boilerplate code that has to be added in several places.